libservice/src/IpAddressNetlinkChecker.cpp

```diff
@@ -131,11 +139,73 @@ static std::optional<IPv4int> getMappedIPv4Addr(const in6_addr& addr) {
 	return ipv4Binary;
 }
 
+IpAddressNetlinkChecker::ipv6Range IpAddressNetlinkChecker::parseIpv6Range(const std::string& range) const {
+	ipv6Range rangeStruct;
+	const auto slashPos{range.find('/')};
+	rangeStruct.ipv6Address = range.substr(0, slashPos);
+	rangeStruct.prefixLength = slashPos != std::string::npos ? std::stoi(range.substr(slashPos + 1)) : 0;
+	return rangeStruct;
+}
+
+bool IpAddressNetlinkChecker::isInRange(const in6_addr& addr, const std::string& range) const {
+	auto rangeStruct{IpAddressNetlinkChecker::parseIpv6Range(range)};
+
+	in6_addr rangeIpv6Address{};
+	inet_pton(AF_INET6, rangeStruct.ipv6Address.c_str(), &rangeIpv6Address);
+
+	// Create mask
+	in6_addr mask{};
+	for (size_t i = 0; i < sizeof(mask.s6_addr); i++) {
+		if (rangeStruct.prefixLength >= 8) {
+			mask.s6_addr[i] = 0xFF;
+			rangeStruct.prefixLength -= 8;
+		} else if (rangeStruct.prefixLength > 0) {
+			mask.s6_addr[i] = (uint8_t)(0xFF << (8 - rangeStruct.prefixLength));
+			rangeStruct.prefixLength = 0;
+		} else {
+			mask.s6_addr[i] = 0;
+		}
+	}
+
+	// Check mask and addr
+	in6_addr andResult{};
+	for (size_t i = 0; i < sizeof(andResult.s6_addr); ++i) {
+		andResult.s6_addr[i] = addr.s6_addr[i] & mask.s6_addr[i];
+	}
+
+	// Compare andResult with IPv6 address of the rangeStruct
+	return memcmp(&andResult, &rangeIpv6Address, sizeof(andResult)) == 0;
+}
+
+bool IpAddressNetlinkChecker::checkSubnet(
+		const in6_addr& addrToCheck, const in6_addr& interfaceIpv6Addr, const in6_addr& interfaceMask) const {
+	const auto s6Addr32ArraySize = sizeof(addrToCheck.s6_addr32) / sizeof(addrToCheck.s6_addr32[0]);
+	for (size_t i = 0; i < s6Addr32ArraySize; ++i) {
+		if ((addrToCheck.s6_addr32[i] & interfaceMask.s6_addr32[i]) != (interfaceIpv6Addr.s6_addr32[i] & interfaceMask.s6_addr32[i])) {
+			return false;
+		}
+	}
+	return true;
+}
+
 bool IpAddressNetlinkChecker::isV6AddressExternal(const in6_addr& addr) const {
```

Comment:
shouldn't we check if address belongs to local bridge as well, as we do for ip v4?

Reply:
Function `getifaddrs()` returns with interfaces also local bridges:

```
lukasz@lukasz-ubuntu:~$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br0 state UP group default qlen 1000
    link/ether 00:15:5d:01:8e:02 brd ff:ff:ff:ff:ff:ff
    inet 10.46.1.2/24 brd 10.46.1.255 scope global noprefixroute eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::55ff:e454:33c0:10fd/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br0 state UP group default qlen 1000
    link/ether 00:15:5d:01:8e:09 brd ff:ff:ff:ff:ff:ff
    inet 192.168.129.115/20 brd 192.168.143.255 scope global dynamic noprefixroute eth1
       valid_lft 70183sec preferred_lft 70183sec
    inet6 fe80::d626:67a1:2a6e:e2f6/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
4: eth2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:15:5d:01:8e:0a brd ff:ff:ff:ff:ff:ff
5: br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 8e:19:27:84:65:bb brd ff:ff:ff:ff:ff:ff
    inet6 2001:db8:1234::1234:1234:1234/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::8c19:27ff:fe84:65bb/64 scope link 
       valid_lft forever preferred_lft forever
lukasz@lukasz-ubuntu:~$ brctl show
bridge name	bridge id		STP enabled	interfaces
br0		8000.8e19278465bb	no		eth0
							eth1
```

```
/home/lukasz/dev/eBPF-Discovery/cmake-build-debug/bin/ebpfdiscoverysrv --log-level=info
2024-05-20 14:50:34.210 [132384] info [ebpfdiscovery] 3 network interfaces have been discovered:
2024-05-20 14:50:34.210 [132384] info [ebpfdiscovery] index: 1, IP addresses: 127.0.0.1
2024-05-20 14:50:34.210 [132384] info [ebpfdiscovery] index: 2, IP addresses: 10.46.1.2
2024-05-20 14:50:34.210 [132384] info [ebpfdiscovery] index: 3, IP addresses: 192.168.129.115
2024-05-20 14:52:10.256 [132384] info [ebpfdiscovery] Detected IPv6 interface: ::1
2024-05-20 14:52:10.256 [132384] info [ebpfdiscovery] Detected IPv6 interface: fe80::55ff:e454:33c0:10fd
2024-05-20 14:52:10.256 [132384] info [ebpfdiscovery] Detected IPv6 interface: fe80::d626:67a1:2a6e:e2f6
2024-05-20 14:52:10.256 [132384] info [ebpfdiscovery] Detected IPv6 interface: 2001:db8:1234::1234:1234:1234
2024-05-20 14:52:10.256 [132384] info [ebpfdiscovery] Detected IPv6 interface: fe80::8c19:27ff:fe84:65bb
{"service":[{"pid":125562,"endpoint":"[::1]/","externalClientsNumber":1}]}
```
(IPv6 interfaces printed only for debug)

Reply:
could we then unify implementation for both ipv4 and ipv6 get rid off direct netlink calls?

Reply:
There will be separate PR for implementation unify for both ipv4 and ipv6 and get rid off direct netlink calls. I will extend test coverage in `test_discovery.py`